code/modules/puddles/puddles.dm

```diff
@@ -0,0 +1,62 @@
+/obj/puddle
+	name = "Puddle"
+	icon = 'icons/obj/puddle.dmi'
+	icon_state = "puddle" // just use error state for now
+	alpha = 0
+
+/obj/puddle/New()
+	SSpuddle.puddlelist += src
+	reagents = new(1500)
+
+/obj/puddle/proc/update()
+	color = mix_color_from_reagents(reagents.reagent_list)
+	var/one = reagents.total_volume / reagents.maximum_volume
+	alpha = 75 + (one * 180)
+	layer = one * 5
+	if(reagents.total_volume == 0)
+		qdel(src)
+		// update layer here
+
+/obj/puddle/proc/spread()
+	var/percent = (reagents.total_volume / reagents.maximum_volume) * 100
+	if(percent < 1)
+		return
+
+	// get list of turfs with and without nearby puddles
+	var/list/nearby_puddles = list()
+	var/list/nearby_empty_turfs = list()
+	for(var/turf/open/O in range(1, src))
+		if(loc == O)  // my home
+			continue
+		var/obj/puddle/P = locate() in O
+		if(P)
+			nearby_puddles |= P
+		else
+			if(isspaceturf(O))
+				continue
+			var/obj/machinery/door/airlock/A = locate() in O
+			if(A && A.density)
+				continue
+			nearby_empty_turfs |= O
```

Comment:
This is pretty horrific.

Reply:
Look at how `/obj/effect/particle_effect/foam` does it in `effects_foam.dm` for fairly sane spreading.